WIP: Schema Translation Test + Nested Support in Query Validation Test

rodesai opened the pull request:
### Description 

Refactor the QueryTranslationTest into EndToEndEngineTest and QueryTranslationTest.
EndToEndEngineTest contains the support classes for describing a test case, piping data
into the Streams app, and validating the result. QueryTranslationTest implements test
discovery for the query validation tests.

Extend EndToEndEngineTest to support arbitrary record value types. Record now stores the value
as Object, and returns a Supplier interface for getting a Serde given a SchemaRegistryClient.

Add support for serializing/deserializing to/from a generic record spec format. Tests that are built
on EndToEndEngineTest can use the generic format to define inputs and expected outputs. This
Is useful for defining nested rows, for example. For now this format is just the json representation
of a row.

Adds a schema translation test (SchemaTranslationTest. This test can be used to validate that
KSQL loads avro data formatted with a given schema correctly. The test is built on the EndToEndEngineTest. Each test case describes an avro schema under test, and an optional
list of expected inputs and outputs. The test creates a source stream that uses the schema and
runs a simple projection of all the contained fields. If no input/output records are specified, random
input records are generated and the expected output should match the input.

### Testing done 
_Describe the testing stategy. Unit and integration tests are expected for any behavior changes._

### Reviewer checklist
- [ ] Ensure docs are updated if necessary. (eg. if a user visible feature is being added or changed).
- [ ] Ensure relevant issues are linked (description should include text like "Fixes #<issue number>")

rodesai committed:
WIP: Schema Translation Test + Nested Support in Query Validation Test

Refactor the QueryTranslationTest into EndToEndEngineTest and QueryTranslationTest.
EndToEndEngineTest contains the support classes for describing a test case, piping data
into the Streams app, and validating the result. QueryTranslationTest implements test
discovery for the query validation tests.

Extend EndToEndEngineTest to support arbitrary record value types. Record now stores the value
as Object, and returns a Supplier interface for getting a Serde given a SchemaRegistryClient.

Add support for serializing/deserializing to/from a generic record spec format. Tests that are built
on EndToEndEngineTest can use the generic format to define inputs and expected outputs. This
Is useful for defining nested rows, for example. For now this format is just the json representation
of a row.

Adds a schema translation test (SchemaTranslationTest. This test can be used to validate that
KSQL loads avro data formatted with a given schema correctly. The test is built on the EndToEndEngineTest. Each test case describes an avro schema under test, and an optional
list of expected inputs and outputs. The test creates a source stream that uses the schema and
runs a simple projection of all the contained fields. If no input/output records are specified, random
input records are generated and the expected output should match the input.

rodesai committed:
fix some merge problems

dguy commented on ksql-engine/src/test/java/io/confluent/ksql/EndToEndEngineTest.java:
```diff
@@ -0,0 +1,554 @@
+import io.confluent.ksql.util.QueryMetadata;
+
+@Ignore
+public class EndToEndEngineTest {
```
I'd much prefer we didn't use inheritance. Composition is usually a better option. I think we should be able to use most of this code as is without the inheritance.

big-andy-coates replied:
I'd _strongly_ agree. Inheritance should be used very sparingly in code. Composition is definitely the way to go here. Make reusable components that tests can pick up and use as needed. Make them as junit `ExternalResource`s for added bling.

rodesai replied:
I agree with the principle. In this case my thinking was that inheriting avoids boilerplate for the core test method in all the children. The hierarchy/functionality is pretty limited in scope (1 externally facing function, 1 level contained to 1 package, each child implements only static methods). Each child is really just an EndToEndEngine test with an optimized config file format. So its unlikely we'd hit the usual encapsulation-violation problems. All that said, the only real benefit is saving some boilerplate. And its not much boilerplate being saved. So I'm fine with changing it.

dguy commented on ksql-engine/src/test/java/io/confluent/ksql/EndToEndEngineTest.java:
```diff
@@ -0,0 +1,554 @@
+    put(StreamsConfig.STATE_DIR_CONFIG, TestUtils.tempDirectory().getPath());
+  }};
+  private final Properties streamsProperties = new Properties();
+  private SchemaRegistryClient schemaRegistryClient = new MockSchemaRegistryClient();
```
can this be final?

rodesai replied:
yeah there are a few other places I need to add final as well.

dguy commented on ksql-engine/src/test/java/io/confluent/ksql/EndToEndEngineTest.java:
```diff
@@ -0,0 +1,554 @@
+      } catch (Exception e) {
+        throw new RuntimeException(e);
+      }
+      new org.apache.avro.Schema.Parser().parse(schemaString);
```
not sure what the purpose of this line is?

rodesai replied:
copypasta

dguy commented on ksql-engine/src/test/java/io/confluent/ksql/EndToEndEngineTest.java:
```diff
@@ -0,0 +1,554 @@
+      } catch (Exception e) {
+        throw new RuntimeException(e);
+      }
+      new org.apache.avro.Schema.Parser().parse(schemaString);
```
ditto here

dguy commented on ksql-engine/src/test/java/io/confluent/ksql/EndToEndEngineTest.java:
```diff
@@ -0,0 +1,554 @@
+  }
+
+  protected static class Record {
+    private final String topic;
```
Wondering if this should be `Topic` now? and the `SerdeSupplier` could go on the `Topic`?

rodesai replied:
Yes this makes sense.

rodesai committed:
Apply review feedback

rodesai committed:
Remove copied parse calls

dguy commented on ksql-engine/src/test/java/io/confluent/ksql/EndToEndEngineTest.java:
```diff
@@ -0,0 +1,531 @@
+import io.confluent.ksql.util.KsqlConfig;
+import io.confluent.ksql.util.QueryMetadata;
+
+class EndToEndEngineTest {
```
Seeing as this is not a test anymore maybe rename it to something else? `EndToEndEngineTestUtils` or something?

dguy commented on ksql-engine/src/test/java/io/confluent/ksql/EndToEndEngineTest.java:
```diff
@@ -0,0 +1,531 @@
+      } catch (Exception e) {
+        throw new RuntimeException(e);
+      }
+      return avroToValueSpec(
```
nit: you could move this up into the `try{..}catch` and then `schemaString` can be initialized where it is defined

rodesai replied:
I'd prefer to minimize the code under the catch-all to avoid rewrapping exceptions as much as possible.

dguy commented on ksql-engine/src/test/java/io/confluent/ksql/EndToEndEngineTest.java:
```diff
@@ -0,0 +1,531 @@
+      } catch (Exception e) {
+        throw new RuntimeException(e);
+      }
+      final Object avroObject = valueSpecToAvro(
```
as above

dguy commented on ksql-engine/src/test/java/io/confluent/ksql/EndToEndEngineTest.java:
```diff
@@ -0,0 +1,531 @@
+          r -> testDriver.pipeInput(
+              new ConsumerRecordFactory<>(
+                  Serdes.String().serializer(),
+                  r.topic.getSerdeSupplier().getSerializer(schemaRegistryClient)
```
maybe add `getSerializer()` to `Topic`, i.e., to avoid the chaining

dguy commented on ksql-engine/src/test/java/io/confluent/ksql/EndToEndEngineTest.java:
```diff
@@ -0,0 +1,531 @@
+    }
+  }
+
+  private static TopologyTestDriver buildStreamsTopology(final Query query, final KsqlEngine ksqlEngine,
```
nit: One param per line of split across multiple

dguy commented on ksql-engine/src/test/java/io/confluent/ksql/EndToEndEngineTest.java:
```diff
@@ -0,0 +1,531 @@
+  static void shouldBuildAndExecuteQuery(Query query) {
+    final MetaStore metaStore = new MetaStoreImpl(new InternalFunctionRegistry());
+    final Map<String, Object> config = new HashMap<String, Object>() {{
+      put(ConsumerConfig.BOOTSTRAP_SERVERS_CONFIG, "localhost:9092");
```
I think this should be the address of the embedded kafka?

rodesai replied:
The test doesn't actually run any broker. The config is just there to placate streams and admin client.

dguy reviewed (commented):
Thanks @rodesai, left a few more comments

big-andy-coates committed:
#1396 Fix regression issue with string concat using '+' operator. (#1406)

* #1396 Fix regression issue with string concat using '+' operator.

rodesai committed:
Add struct support for avro by using Connect Converter API (#1400)

Add struct support for avro by using Connect Converter API

Now that we're moving towards using Connect's data types as KSQL's internal
representation for records, we can use the Connect Converters to (de)serialize
records. This patch implements a generic Connector-based serializers/deserializer,
and switches KsqlAvroTopicSerde to use that serializer/deserializer, configured
with an Avro converter.

We still require a thin translation layer in KSQL to handle case- insensitivity
, and some of our current limits (e.g. no support for binary fields). This translation
layer is implemented by ConnectSchemaTranslator and ConnectDataTranslator.

Finally, Connect Structs validate that fields are optional when assigning null. This
patch includes a change to make all struct fields optional. In order to switch to using
Struct instead of GenericRow we should just make all our fields optional, but that can
come in a later patch.

rodesai committed:
Return schema description API entities as JSON objects (#1362)

* Return api schema description as a JSON object instead of a string

Switch field schemas returned in stream/query descriptions to JSON objects
from strings. Now that we're adding support for struct fields, each field
can have a complex structure. We should return this as a structured object
rather than trying to serialize the whole schema into a single string field.

dguy committed:
Create and load Java based UDFs from an ext directory (#1366)

Adds the ability to deploy udf jars in an ext directory. Udfs are loaded from the jars by searching for the anotations @UdfDescription and @Udf. Each method (in a class with @UdfDescription) marked with @Udf is compiled into a class that can be called via the existing function call code. Each UDF gets it's own child first classloader so that they are free to include versions of jars that are different than those provided by ksql.

Classes can be "Blacklisted" from being called from a UDF, i.e., so they can't call System.exit(), for example.

Have also changed the SubstringKudf to this new form and it is now loaded in the same way, but from the KSQL classpath.

#1421

miguno committed:
Add year 2018 to the copyright info in the CLI banner (#1415)

miguno committed:
Add missing comma to fix broken REST API example (#1414)

hjafarpour committed:
Added query rewrite feature using AST. (#1410)

* Added query rewrite feature using AST.

* Applied review feedback.

* Forgot one final!

* Fixed line length issue.

* Applied more review feedback.

* Added some comments.

hjafarpour committed:
Minor style change in comment. (#1428)

dguy committed:
Validate that function names are valid java identifiers (#1416)

Validate that function names are:

not a java reserved word
not a KSQL reserved word
not null or empty
is a valid java identifier

#1421

JimGalasyn committed:
DOCS-415: New topic on KSQL custom test data (#1412)

* A few tweaks

* Add test-data topic to index

* Add semicolon to persistent query

* Remove schema registration section

JimGalasyn committed:
DOCS-415: New topic on KSQL custom test data (#1412)

* A few tweaks

* Add test-data topic to index

* Add semicolon to persistent query

* Remove schema registration section

JimGalasyn committed:
DOCS-415: New topic on KSQL custom test data (#1412)

* A few tweaks

* Add test-data topic to index

* Add semicolon to persistent query

* Remove schema registration section

JimGalasyn committed:
DOCS-415: New topic on KSQL custom test data (#1412)

* A few tweaks

* Add test-data topic to index

* Add semicolon to persistent query

* Remove schema registration section

dguy committed:
Collect metrics for UDFs (#1423)

Adds collection of metrics for UDFs if the config ksql.udf.collect.metrics is set to true. The default is false.

As it stands the metrics are global (rather than per query) as they are intended for testing purposes rather than production, i.e., i think the overhead may be too much to collect them on every invocation.

#1421

rodesai committed:
Address more review feedback

maxzheng committed:
Merge branch '4.1.0-post' into 4.1.1-post

maxzheng committed:
Merge branch '4.1.1-post' into 4.1.x

maxzheng committed:
Merge branch '4.1.x' into 5.0.x

rodesai committed:
More review feedback

dguy committed:
UDF tidy ups (#1429)

* UDF tidy ups
#1421

dguy committed:
Throw exception when calling FunctionRegistry#getUdfFactory(name) if function doesn't exist (#1426)

* Throw exception when calling FunctionRegistry#getUdfFactory(name) if the function doesn't exist
#1421

dguy committed:
update packaging to include ext dir for udfs (#1433)

big-andy-coates committed:
Prefixed output topics (#1445)

* Support setting properties in QueryTranslationTest

* Code refactor / warnings fixed.

* Add new KsqlConig.KSQL_OUTPUT_TOPIC_NAME_PREFIX_CONFIG = "ksql.output.topic.name.prefix";

Which allows users to specifiy a prefix to be applied to the names of any ouput topics that are created from statements without an explicit topic name supplied.

e.g.

Given `ksql.output.topic.name.prefix=my-first-prefix`

```
CREATE STREAM S AS ...  -- will use Kafka topic with name 'my-first-prefix-S'
CREATE STREAM S WITH(KAFKA_TOPIC='foo') AS ... -- will use Kafka topic with name 'foo'
```

This is useful for users of interactive clusters wanting to set up ACLs on a prefixed resource pattern in Kafka cluster to allow users to create output topics. e.g. in the above example, an ACL could be set up to allow KSQL to create/write to any topic prefixed with 'my-first-prefix-'.

rodesai committed:
Add complex fields to the basic avro tets

big-andy-coates committed:
Simplified Security (#1442)

* Update docs to make use of ACLs on prefixed resource patterns for AK clusters 2.0 and beyond

hjafarpour committed:
Updated AST builder to handle dot notation. (#1430)

* Updated AST builder to handle dot notation.

* Removed some duplicate code by making some private constructors public.

* Applied feedback.

* Going back to the previous implementation with a minor twist.

* Minor changes.

* Minor changes.

JimGalasyn committed:
Fix munged KSQL query code block (#1451)

JimGalasyn committed:
Merge branch '4.1.0-post' into 4.1.1-post

JimGalasyn committed:
Merge branch '4.1.1-post' into 4.1.x

JimGalasyn committed:
Merge branch '4.1.x' into 5.0.x

apurvam committed:
Add new join types to KSQL (#1417)

This patch adds LEFT, OUTER, and INNER stream-stream and table-table joins. Additionally it adds an INNER stream-table join.

Majority of the logic is in JoinNode where a Joiner class hierarchy is introduced to do the heavy lifting of executing the right join. This hierarchy maximizes code reuse across the join types.

Lot's of redundant grammar constructs were removed as part of this patch. In particular, we had lexical support far JOIN USING instead of JOIN ON. We had support for CROSS joins in the grammar. These have been removed.

Fixes #731. Fixes #760 . Fixes #1380 .

dguy committed:
Support List/Show Functions (UDFs) and Describe Function <name> (#1435)

Add new syntax LIST FUNCTIONS; , SHOW FUNCTIONS;, and DESCRIBE FUNCTION <name>;.
Will show a list of all UDFs (note doesn't include aggregations at the moment), i.e.,

 Function Name     
-------------------
 TIMESTAMPTOSTRING 
 TRIM              
 SUBSTRING         
 EXTRACTJSONFIELD  
 RANDOM            
 ROUND             
 CONCAT            
 ARRAYCONTAINS     
 UCASE             
 GEO_DISTANCE      
 FLOOR             
 ABS               
 LCASE             
 LEN               
 IFNULL            
 STRINGTOTIMESTAMP 
 CEIL              
-------------------
DESCRIBE FUNCTION substring;

Name        : SUBSTRING
Author      : Confluent
Version     : 
Overview    : returns a substring of the passed in value
Variations  : 

	Arguments   : VARCHAR, INT
	Returns     : VARCHAR
	Description : Returns a string that is a substring of this string. The substring begins with the character at the specified startIndex and extends to the end of this string.

	Arguments   : VARCHAR, INT, INT
	Returns     : VARCHAR
	Description : Returns a string that is a substring of this string. The substring begins with the character at the specified startIndex and extends to the character at endIndex -1.

apurvam committed:
Rename SPAN to WITHIN everywhere (#1455)

ijuma committed:
Handle rename of KafkaConsumerService to ConsumerService (#1463)

Note that this class is internal and we should migrate to the
public AdminClient as soon as possible.

big-andy-coates committed:
Fix some doc formatting issues. (#1461)

* Fix some doc formatting issues.

dguy committed:
Add config for ext dir (#1448)

* externalize ext-dir config

rodesai committed:
Use a string for formatted time field in schema (#1470)

dguy committed:
create module for UDF annotations to minimize dependencies when creating them (#1449)

dguy committed:
Simple Security manager for UDFS (#1457)

* add security manager that is used by PluggableUdf
* only stop exit and forking of processes

dguy committed:
add jar path to description of functions (#1458)

* add jar path to description of functions

dguy committed:
support null literal in UDFs #1462 (#1465)

* support null literal in UDFs

apurvam committed:
Renable the build of the cp-ksql-server docker image (#1471)

hjafarpour committed:
Struct end to end (#1453)

* PR2: struct rewrite

* PR3-End-To-End struct support

* Applied Andy and Damian's feedback.

* Applied Rohan's feedback.

* More changes based on Andy's feedback.

* More improvements based on reveiws.

* Applied final review suggestions.

rodesai committed:
Upgrade handling step 1: clean up KSQL configs (#1390)

* Upgrade handling step 1: clean up KSQL configs

This patch does some housekeeping of our KSQL configs:
- Make KsqlConfig read-only. Having a read-only object makes the code that
  uses it much easier to reason about. Previously we had a put interface used
  to tweak config at process start, and to pass around data between internal
  classes. Tweaking configs can be solved by preprocessing the properties or
  using the clone-with-overwrites interface before passing the KsqlConfig object
  to all the classes that need it. The latter usage (passing around data between
  internal classes) is just a bad way to pass data around. I was able to clean
  up all the places we were doing this, except for the timestamp column index case
  where different plan nodes use the ksql.timestamp.column.index config to pass
  around the column used to extract timestamps. For the time being I left in an
  interface to handle this case specifically ([gs]etKsqlTimestampColumnIndex).
- Filter the passed in configs for each subsystem (for now just streams).
- Use fully-qualified config names in listings. The use of un-qualified names is
  just a convenience when setting the configs.
- Get rid of the generic get() interface and use the typed interfaces provided
  by AbstractConfig.
- Extend the "list properties" response to indicate which properties are overridden.
- Print out all the configs in the CLI in sorted order

* Rebase and fix some issues

* address review feedback

* fix checkstyle

* Fix broken tests

dguy committed:
Support java based UDAFS (#1467)

* Suppor java based UDAFS
#1421

big-andy-coates committed:
Output more info to help with debugging system tests. (#1494)

hjafarpour committed:
Use ':' for struct field reference. (#1478)

* Use ':' for struct field reference.

* Switched to `->` from `:` for struct field reference.
Removed the unused Lambda expression from the grammar since it was also using `->`.

* Added ambiguity check test.
Applied Andy's feedback.
Fixed a minor case sensitivity bug.

* Applied feedback from Rohand and Andy.

rodesai committed:
Fix rebase problems
